sled-hardware/src/disk.rs

```diff
@@ -4,15 +4,22 @@
 
 use camino::{Utf8Path, Utf8PathBuf};
 use illumos_utils::fstyp::Fstyp;
+use illumos_utils::zfs::EncryptionDetails;
+use illumos_utils::zfs::Keypath;
 use illumos_utils::zfs::Mountpoint;
 use illumos_utils::zfs::Zfs;
 use illumos_utils::zpool::Zpool;
 use illumos_utils::zpool::ZpoolKind;
 use illumos_utils::zpool::ZpoolName;
+use key_manager::StorageKeyRequester;
 use slog::Logger;
 use slog::{info, warn};
+use tokio::fs::{remove_file, File};
+use tokio::io::{AsyncSeekExt, AsyncWriteExt, SeekFrom};
 use uuid::Uuid;
 
+pub const KEYPATH_ROOT: &str = "/var/run/oxide/";
```

Comment:
Reading below: Seems like this would be a significant problem if this was lost between boots?

Reply:
This is very much intentional. We want to limit the existence of the keyfile to the minimum amount of time necessary to either create the encrypted filesystem or to load the keys so we can mount it. After that we delete it. This key is actually just a wrapper around the internal ZFS key and we want to limit any extra mechanisms to access that.

The reason this works is that the `KeyManager` will re-derive the key from the secret for the given epoch on demand, and we rewrite the key file before we need it. I've tested this on cold-boot a number of times and it works. Since up until recently we didn't have persistent M.2 data, I had to redeploy after rebooting a sled and run RSS again. But since the zpools and datasets already exist, they just get mounted after the key is recreated. I verified they are not re-created by writing out a file to the encrypted dataset that I then read-back after reboot.